DAOS-9756 vos: Track latest aggregatable write

jolivier23 opened the pull request:
Move filtering of entries to the filter callback to avoid ilog scans if entries can be
trivially skipped.

Save the upper 32-bits of the HLC (or for RDB and other entities, the lower 32-bits
of the epoch) for the last aggregatable write to the tree feats so we can filter entries
that don't need to be aggregated from the scan at container, object, dkey, and akey
level.  For now, the last aggregatable write is defined as an object or key punch or
the 2nd write to a single value or array.  This can be further refined for array values
in the future to only update it if there are adjacent or overlapping extents.

This allows us to easily skip unchanged entries or entries that will be noops for
aggregation.

Signed-off-by: Jeff Olivier <jeffrey.v.olivier@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/1/execution/node/727/log

daosbuild1 commented:
Test stage Unit Test with memcheck completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/1/execution/node/755/log

NiuYawei commented on src/vos/vos_io.c:
```diff
@@ -1659,6 +1661,71 @@ akey_update_recx(daos_handle_t toh, uint32_t pm_ver, daos_recx_t *recx,
 }
 
+int
+vos_key_flags_set(struct umem_instance *umm, struct vos_krec_df *krec)
```
This function name looks bit confusing to me, it's trying to mark the item "aggregate-able" but not for setting a common feat flag, so vos_key_mark_agg() or something else looks better to me.

jolivier23 replied:
Yeah, i can fix that

NiuYawei commented on src/vos/vos_io.c:
```diff
@@ -1659,6 +1661,71 @@ akey_update_recx(daos_handle_t toh, uint32_t pm_ver, daos_recx_t *recx,
+		return 0;
+
+	if ((bmap & KREC_BF_AGG_NEEDED) == 0)
+		bmap |= KREC_BF_AGG_NEEDED;
```
I don't see why the KREC_BF_AGG_FLAG isn't cleared in such case.

jolivier23 replied:
It should never be set in such case.

NiuYawei replied:
oh, right. probably adding an assert here?

NiuYawei commented on src/vos/vos_obj.c:
```diff
@@ -311,8 +311,11 @@ key_punch(struct vos_object *obj, daos_epoch_t epoch, daos_epoch_t bound,
-	if (rc != 1)
+	if (rc != 1) {
+		/** Otherwise, key_tree_punch will handle it */
+		rc = vos_key_flags_set(vos_cont2umm(obj->obj_cont), krec);
```
Should akey being marked when it's a akey punch?

jolivier23 replied:
It is handled inside of key_tree_punch

NiuYawei commented on src/vos/vos_obj.c:
```diff
@@ -499,6 +502,10 @@ vos_obj_punch(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
 
+			if (rc == 0)
+				rc = vos_obj_flags_set(vos_cont2umm(cont), &obj->obj_df->vo_tree,
+						       &cont->vc_cont_df->cd_obj_root);
```
Why container isn't marked?

jolivier23 replied:
You mean pool?   Container is marked by setting bits in the object tree root.

NiuYawei replied:
Yeah, I meant container, I was thinking vos_obj_flags_set() only update flags for object.

NiuYawei commented on src/vos/vos_obj.c:
```diff
@@ -1928,26 +1937,45 @@ vos_obj_iter_pre_aggregate(daos_handle_t ih)
+			if (!punched && oiter->it_punched.pr_epc == 0)
+				return 2;
+		} else {
+			bmap |= KREC_BF_AGG_FLAG;
```
Introduce an extra transaction to for each key to set KREC_BF_AGG_FLAG flag looks overkilled to me.

jolivier23 replied:
how would you handle it otherwise?

NiuYawei replied:
Like tolerating a full scan after server restart and update flags without persistence guarantee. I'm afraid that flushing and invalidating CPU cache when iterating each entry could impact performance, but let's evaluate the performance first then make decision.

jolivier23 replied:
Doesn't cache invalidation only happen with Apache Pass?  I thought they were fixing that in later renditions.

NiuYawei commented on src/vos/vos_aggregate.c:
```diff
@@ -356,11 +363,18 @@ vos_agg_dkey(daos_handle_t ih, vos_iter_entry_t *entry,
 				inc_agg_counter(cont, VOS_ITER_DKEY, AGG_OP_DEL);
 				return 0;
+			} else if (rc == 2) {
+				/** Entry is skipped removed the key so let's reprobe */
```
If the aggregation is a full scan for snapshot deletion (or after rebuild done), we can't skip it here.

jolivier23 replied:
I understand why we can't skip for snapshot deletion.  How do we check for that?   Why rebuild?  If we do rebuild, won't it set the appropriate flags?

NiuYawei replied:
The need_aggregate() above actually checked this by checking the "VOS_AGG_FL_FORCE_SCAN" flag, I'm wondering if should move these feat flags check into need_aggregate()?

We treat rebuild in a similar way to snapshot, rebuild will create a fence epoch (like snapshot epoch) to avoid aggregation spanning the fence, once rebuild is done, the fence epoch will be deleted and a full scan will be triggered.

jolivier23 replied:
I wonder whether we should just disable aggregation during rebuild

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

NiuYawei reviewed (commented):
Updating VOS_TREE_AGG_FLAG for each key in aggregation looks overkilled to me.

Maybe we can make a tradeoff here, updating these flags without persistence guarantee (no PMDK transactions), and tolerate a full scan after server restart, what do you think?

jolivier23 commented on src/vos/vos_io.c:
```diff
@@ -1659,6 +1661,71 @@ akey_update_recx(daos_handle_t toh, uint32_t pm_ver, daos_recx_t *recx,
+	uint8_t	bmap;
+
+	bmap = krec->kr_bmap;
+	if ((bmap & KREC_BF_AGG_OPT) == 0)
```
Yes, it's fully backward compatible.  If the flag isn't set, it's created by an older version.

NiuYawei replied:
Ok, I was thinking we don't need to consider downgrade support.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

NiuYawei commented on src/vos/vos_obj.c:
```diff
@@ -1928,26 +1938,45 @@ vos_obj_iter_pre_aggregate(daos_handle_t ih)
+
+	bmap = krec->kr_bmap;
+
+	if (bmap & KREC_BF_AGG_OPT && !oiter->it_iter.it_for_discard) {
```
Will it cleaner and easier to follow if we do the feat flags check in need_aggregate()?

jolivier23 replied:
Probably.  I think it probably needs a bit of refactoring.   It's tricky to do this because the data is embedded in the tree root/key record and we don't export this via the iterator.

NiuYawei commented on src/vos/vos_obj.c:
```diff
@@ -1928,26 +1938,45 @@ vos_obj_iter_pre_aggregate(daos_handle_t ih)
+
+	if (bmap & KREC_BF_AGG_OPT && !oiter->it_iter.it_for_discard) {
+		if ((bmap & KREC_BF_AGG_NEEDED) == 0) {
+			if (!punched && oiter->it_punched.pr_epc == 0)
```
What's this "pr_epc == 0" for? could you explain it a bit?

jolivier23 replied:
When we punch something, the subtree flags don't get updated.   So we have to scan a tree if it's punched at a higher level.   This covers that case.  If pr_epc is not zero, it means there is an upper level punch.

jolivier23 replied:
If the key is punched at a higher level, this will be set.  And since we don't scan subtrees and mark them as agg needed, we simply always run aggregation in this case.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Unit Test with memcheck completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/4/execution/node/648/log

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/4/execution/node/699/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

NiuYawei commented on src/vos/vos_io.c:
```diff
@@ -1659,6 +1661,70 @@ akey_update_recx(daos_handle_t toh, uint32_t pm_ver, daos_recx_t *recx,
 }
 
+int
+vos_key_mark_agg(struct umem_instance *umm, struct vos_krec_df *krec)
```
Just curious, why the flags for key are not stored in the btr or evt root feats like object/container? It looks to me not much difference but may save few bits for the 'kr_bmap'.

jolivier23 replied:
I suppose it could be done this way.  Let me look into it.   The main reason was for simplicity but yeah, kr_bmap is limited

NiuYawei commented on src/vos/vos_obj.c:
```diff
@@ -1928,26 +1938,44 @@ vos_obj_iter_pre_aggregate(daos_handle_t ih)
+	bmap = krec->kr_bmap;
+	if (oiter->it_iter.it_for_purge && bmap & KREC_BF_AGG_OPT) {
+		if ((bmap & KREC_BF_AGG_NEEDED) == 0) {
+			if (!full_scan && !punched && oiter->it_punched.pr_epc == 0)
```
It would be great to add a comment to explain the 'pr_epc == 0' here, and I still don't quite see: if the key is punched at a higher level, won't the whole tree already being moved to GC?

NiuYawei replied:
Oh, nm. I see that the tree won't be moved to GC if it's still referenced by snapshots. Some comments could be helpful, thanks!

jolivier23 replied:
Added a comment

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

jolivier23 commented:
I've updated this to avoid snapshot when it's not really necessary (e.g. rollback not really needed).   I've also updated it to track whether we skipped anything (or may have skipped anything) so we don't clear the NEEDED flag prematurely.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

jolivier23 commented:
well, it appears that something may not be working right...now if I can figure out how to run those tests

NiuYawei commented on src/vos/vos_obj.c:
```diff
@@ -947,6 +958,9 @@ key_iter_probe(struct vos_obj_iter *oiter, daos_anchor_t *anchor)
 {
 	int	rc;
 
+	/** If we are reprobing, assume we've skipped something */
```
I think reprobing will always happen since iteration has to yield from time to time, that means the it_skipped will always be set and the 'need_agg" flag will never be cleared?

jolivier23 replied:
Maybe we don't need this.  If we instead track whether we set the AGG_FLAG, we can avoid clearing NEEDED flag if it's a new write

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

Nasf-Fan commented on src/vos/vos_io.c:
```diff
@@ -1685,12 +1752,17 @@ akey_update(struct vos_io_context *ioc, uint32_t pm_ver, daos_handle_t ak_toh,
 			      &iod->iod_name, flags, DAOS_INTENT_UPDATE,
 			      &krec, &toh, ioc->ic_ts_set);
-	if (rc != 0) {
+	if (rc != 0 && rc != 1) {
```
"if (rc < 0)"?

Nasf-Fan commented on src/vos/vos_obj.c:
```diff
@@ -1929,8 +1953,38 @@ vos_obj_iter_pre_aggregate(daos_handle_t ih)
+			feats = dbtree_feats_get(&krec->kr_btr);
+		else
+			feats = evt_feats_get(&krec->kr_evt);
+		if (oiter->it_iter.it_for_purge && feats & VOS_TREE_AGG_OPT) {
```
"oiter->it_iter.it_for_purge" has been checked at line 1958.

Nasf-Fan commented on src/vos/vos_tree.c:
```diff
@@ -1014,7 +1013,15 @@ key_tree_prepare(struct vos_object *obj, daos_handle_t toh,
+	/* As a first cut for aggregation detection, just return 1 if it's not the first update.
+	 * This can be improved upon for evtree to take into account actual extent overlap but
+	 * this is a simpler solution for now.
+	 */
```
For non-created case, we will return 1, then for "!(flags & SUBTR_CREATE)", it must not be the first update, then we need to return 1 instead of 0, right?

jolivier23 replied:
If SUBTR_CREATE isn't set, it's not an update

jolivier23 commented on src/common/btree.c:
```diff
@@ -2184,6 +2184,52 @@ dbtree_update(daos_handle_t toh, d_iov_t *key, d_iov_t *val)
+
+#ifdef DAOS_PMEM_BUILD
+	if (DAOS_ON_VALGRIND) {
+		if (!in_tx) {
```
Yes

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/15/execution/node/358/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/15/execution/node/326/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/15/execution/node/355/log

daosbuild1 commented:
Test stage Build RPM on CentOS 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/15/execution/node/369/log

daosbuild1 commented on src/include/daos/btree.h:
```diff
@@ -545,6 +553,13 @@ int  dbtree_delete(daos_handle_t toh, dbtree_probe_opc_t opc,
+{
+	return root->tr_feats;
+}
+int  dbtree_feats_set(struct btr_root *root, struct umem_instance *umm, uint64_t feats, bool in_tx);
```
```suggestion
}

```

daosbuild1 commented on src/vos/vos_obj.c:
```diff
@@ -1960,8 +1982,38 @@ vos_obj_iter_pre_aggregate(daos_handle_t ih)
 
 	if (!vos_ilog_is_punched(vos_cont2hdl(obj->obj_cont), &krec->kr_ilog, &oiter->it_epr,
-				 &oiter->it_punched, &oiter->it_ilog_info))
+				      &oiter->it_punched, &oiter->it_ilog_info)) {
```
```suggestion
				 &oiter->it_punched, &oiter->it_ilog_info)) {
```

daosbuild1 commented on src/include/daos/btree.h:
```diff
@@ -545,6 +553,13 @@ int  dbtree_delete(daos_handle_t toh, dbtree_probe_opc_t opc,
+{
+	return root->tr_feats;
+}
+int  dbtree_feats_set(struct btr_root *root, struct umem_instance *umm, uint64_t feats);
```
```suggestion
}

```

daosbuild1 replied:
```suggestion
}

```

daosbuild1 replied:
```suggestion
}

```

daosbuild1 replied:
```suggestion
}

```

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/17/execution/node/113/log

daosbuild1 commented on src/vos/vos_aggregate.c:
```diff
@@ -306,26 +351,7 @@ vos_agg_dkey(daos_handle_t ih, vos_iter_entry_t *entry,
-		*acts |= VOS_ITER_CB_SKIP;
-		inc_agg_counter(cont, VOS_ITER_DKEY, AGG_OP_SKIP);
-	}
+	inc_agg_counter(cont, VOS_ITER_DKEY, AGG_OP_SCAN);
```
```suggestion
	struct vos_container	*cont = vos_hdl2cont(agg_param->ap_coh);

```

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/18/execution/node/98/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/19/execution/node/106/log

daosbuild1 commented on src/vos/vos_aggregate.c:
```diff
@@ -306,26 +352,7 @@ vos_agg_dkey(daos_handle_t ih, vos_iter_entry_t *entry,
-		*acts |= VOS_ITER_CB_SKIP;
-		inc_agg_counter(cont, VOS_ITER_DKEY, AGG_OP_SKIP);
-	}
+	inc_agg_counter(cont, VOS_ITER_DKEY, AGG_OP_SCAN);
```
```suggestion
	struct vos_container	*cont = vos_hdl2cont(agg_param->ap_coh);

```

daosbuild1 replied:
```suggestion
	struct vos_container	*cont = vos_hdl2cont(agg_param->ap_coh);

```

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/20/execution/node/98/log

jolivier23 commented on src/vos/evtree.c:
```diff
@@ -2465,6 +2465,7 @@ evt_ent_array_fill(struct evt_context *tcx, enum evt_find_opc find_opc,
 				V_TRACE(DB_TRACE, "Filtered "DF_RECT" filter=("
 					DF_FILTER")\n", DP_RECT(&rtmp),
 					DP_FILTER(filter));
+				tcx->tc_iter.it_skipped = 1;
```
This should be fixed.  It's only valid for sorted iterator and, therefore, EVT_FIND_ALL

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/21/execution/node/107/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

jolivier23 commented on src/vos/vos_obj_index.c:
```diff
@@ -772,10 +844,24 @@ oi_iter_aggregate(daos_handle_t ih, bool range_discard)
+			invisible = true;
+			rc = 0;
+		}
+		if (iter->it_for_purge) {
```
Because this is only for aggregation, not discard.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

jolivier23 commented on src/vos/vos_obj.c:
```diff
@@ -2029,10 +2142,32 @@ vos_obj_iter_aggregate(daos_handle_t ih, bool range_discard)
+			 *  Either way, we can clear the aggregation flag.
+			 */
+			feats = feats & ~VOS_TREE_AGG_FLAG;
+			if (!oiter->it_iter.it_skipped)
```
I suppose if we use the upper bits of the timestamp, we don't have to worry about wrap.  And we can even use more than 32 because we have about 60 feature bits to spare.   Perhaps we could just use the physical clock component of the HLC or upper 46 bits and store it in the same place I'm storing this flag.   And it could retain the aggregatable meaning so rather than last write, it's last aggregatable write

daosbuild1 commented on src/include/daos_srv/evtree.h:
```diff
@@ -282,11 +282,26 @@ enum evt_feats {
+	/** Timestamp mask for last aggregatable write */
+	EVT_FEAT_AGG_TIME_MASK		= (0xffffffffULL << 30),
 };
+D_CASSERT((EVT_FEAT_AGG_OPT & EVT_FEAT_AGG_TIME_MASK) == 0);
```
```suggestion
};

```

daosbuild1 replied:
```suggestion
};

```

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/26/execution/node/113/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/26/execution/node/396/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/26/execution/node/346/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/26/execution/node/382/log

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/26/execution/node/758/log

daosbuild1 commented:
Test stage Unit Test with memcheck completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/26/execution/node/772/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build RPM on CentOS 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/28/execution/node/379/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/28/execution/node/370/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8348/28/execution/node/365/log

NiuYawei commented on src/vos/vos_internal.h:
```diff
@@ -332,9 +332,56 @@ struct vos_dtx_cmt_ent {
+		extra_flag = VOS_TREE_AGG_HLC;
+	} else {
+		epoch &= 0x0ffffffffULL;
+		epoch = (epoch << 29);
```
Is this for the integer epoch used in tests (or some other standalone VOS cases)? then I'm wondering how to compare it with HAE which is represented by HLC.

I'm wondering if we should always set the agg_time as maximum value in such case? so that the VOS_TREE_AGG_HLC bit could be saved?

NiuYawei replied:
Oh, never mind, I forgot that HAE is from epoch but not HLC.

NiuYawei commented on src/vos/vos_obj.c:
```diff
@@ -2029,10 +2106,12 @@ vos_obj_iter_aggregate(daos_handle_t ih, bool range_discard)
+			/* Key no longer exists at epoch but isn't empty */
+			invisible = true;
+			rc = 0;
+		}
```
This change could be reverted.

NiuYawei commented on src/vos/vos_obj_index.c:
```diff
@@ -777,10 +840,12 @@ oi_iter_aggregate(daos_handle_t ih, bool range_discard)
+			/** ilog isn't visible in range but still has some enrtries */
+			invisible = true;
+			rc = 0;
+		}
```
Ditto, the change could be reverted.

NiuYawei commented on src/vos/vos_obj_index.c:
```diff
@@ -730,7 +792,7 @@ oi_iter_pre_aggregate(daos_handle_t ih)
 
 int
-oi_iter_aggregate(daos_handle_t ih, bool range_discard)
+oi_iter_aggregate(daos_handle_t ih, bool range_discard, uint64_t *skipped)
```
The "skipped" parameter is not used and can be removed.

NiuYawei reviewed (approved):
LGTM. If the patch needs be refreshed, please also fix a typo in my credits patch. (in credits_set() of vos_aggregate.c, the AGG_CREDS_SCAN_SLACK for vac_creds_del should be AGG_CREDS_DEL_SLACK), otherwise, I'll push a patch to fix the typo. Thanks.

ryon-jensen commented on src/vos/evtree.c:
```diff
@@ -2515,8 +2515,9 @@ evt_ent_array_fill(struct evt_context *tcx, enum evt_find_opc find_opc,
 						 intent);
 			/* Skip the unavailable record. */
-			if (rc == ALB_UNAVAILABLE)
+			if (rc == ALB_UNAVAILABLE) {
```
nit - can remove braces
